arbnode/espresso_caff_node.go

```diff
@@ -0,0 +1,553 @@
+	var safeBlockNumber uint64
+	if n.configFetcher().WaitForFinalization {
+		// if we have configured to wait for finalizations, fetch the latest finalized block number.
+		blockNumber, err := n.l1Reader.LatestFinalizedBlockNr(context.Background())
+		safeBlockNumber = blockNumber
+		if err != nil {
+			log.Warn("Error getting finalized block header to check safety tolerance of delayed message", "err", err)
+			return false, err
+		}
+
+	} else if n.configFetcher().WaitForConfirmations {
+		// if we are waiting for block confirmations, get the latest header and subtract the required block depth.
+		latestHeader, err := n.l1Reader.Client().HeaderByNumber(context.Background(), nil)
+		if err != nil {
+			log.Warn("Error getting finalized block header to check safety tolerance of delayed message", "err", err)
+			return false, err
+		}
+		safeBlockNumber = latestHeader.Number.Sub(latestHeader.Number, new(big.Int).SetUint64(n.configFetcher().RequiredBlockDepth)).Uint64()
+
+	} else {
+		// If we haven't configured a safety strategy, every delayed message is valid to include in the nodes state.
+		return true, nil
+	}
+
+	// safeBlockNumber will be popluated from here onwards. Any code paths that don't set the variable, return from the function before they get here.
+	delayCount, err := n.delayedMessageFetcher.getDelayedMessageCountAtBlock(safeBlockNumber)
+	if err != nil {
+		log.Warn("Error getting the delayed message count while checking the delayed messages safety tolerance", "err", err)
+		return false, err
+	}
+	if (message.MessageWithMeta.Message.Header.BlockNumber <= safeBlockNumber) && (message.MessageWithMeta.DelayedMessagesRead <= delayCount) {
+		return true, nil
+	}
+	return false, nil
+
+}
+
+// nextMessage wraps the espressoStreamer.Next() method, to handle producing delayed messages by checking they are within the nodes safety tolerance.
+// Returns:
+//   - MessageWithMetadataAndPos: A message, delayed or normally sequenced, that is for the next position in the chain.
+//   - error: If any error is encountered during this function it is propegated to the caller.
+//
+// Semantics:
+//
+//	This function will either produce a message, or an error. When an error is produced, the messageWithMetadataAndPos will be nil.
+//	If the message is populated, the error will be nil.
+func (n *EspressoCaffNode) nextMessage() (*espressostreamer.MessageWithMetadataAndPos, error) {
+	messageWithMetadataAndPos, err := n.espressoStreamer.Next()
+	if err != nil {
+		return nil, err
+	}
+
+	if messageWithMetadataAndPos == nil {
+		return nil, nil
+	}
+
+	if messageWithMetadataAndPos.MessageWithMeta.DelayedMessagesRead == n.nextDelayedCount+1 {
+		log.Info("Getting delayed message", "nextDelayedCount", n.nextDelayedCount)
+		// If this is delayed message, we need to get the message from L1
+		// and replace the message in the messageWithMetadataAndPos
+		message, err := n.delayedMessageFetcher.getDelayedMessage(n.nextDelayedCount)
```

Comment:
Shouldn't we be fetching the delayed message at `messageWithMetadataAndPos.MessageWithMeta.DelayedMessagesRead`, not `n.nextDelayedCount` if the message has a `DelayedMessagesRead` of 3 shouldn't we query the `DelayedMessageFetcher` for index 3?

Reply:
yea you are right, it should be `messageWithMetadataAndPos.MessageWithMeta.DelayedMessagesRead`

Reply:
After discussing about this more, the current logic is correct because we need the index and index is always one less than Delayed messages read